File: sdk/kotlin/src/main/kotlin/app/shadowboy/netplay/sdk/json/NetplayJson.kt
package app.shadowboy.netplay.sdk.json

import app.shadowboy.netplay.sdk.protocol.ClientMessage
import app.shadowboy.netplay.sdk.protocol.ServerMessage
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

public object NetplayJson {
    public val format: Json = Json {
        classDiscriminator = "type"
        encodeDefaults = true
        explicitNulls = false
        ignoreUnknownKeys = true
    }

    public fun encodeClientMessage(message: ClientMessage): String =
        format.encodeToString<ClientMessage>(message)

    public fun decodeServerMessage(payload: String): ServerMessage =
        format.decodeFromString<ServerMessage>(payload)
}

File: sdk/typescript/src/json/netplayJson.ts
import type { ClientMessage, ServerMessage } from "../protocol/messages.ts";

const serverMessageTypes = new Set([
  "roomJoined",
  "roomStateChanged",
  "pong",
  "startSession",
  "inputFrame",
  "linkCablePacket",
  "snapshotChunk",
  "snapshotComplete",
  "sessionPauseScheduled",
  "sessionPauseUpdated",
  "sessionResumeScheduled",
  "compatibilityRequested",
  "recoveryStarted",
  "playerReconnected",
  "playerExited",
  "voiceTokenRefreshed",
  "recoveryResyncRequired",
  "serverFrame",
  "stateHashMismatch",
  "inputDelayChanged",
  "heartbeatAck",
  "error",
]);

export function encodeClientMessage(message: ClientMessage): string {
  return JSON.stringify(message);
}

export function decodeServerMessage(payload: string): ServerMessage {
  const parsed = JSON.parse(payload) as unknown;
  if (!isObject(parsed) || typeof parsed.type !== "string") {
    throw new Error("Netplay server message is missing a type.");
  }
  if (!serverMessageTypes.has(parsed.type)) {
    throw new Error(`Unknown netplay server message type: ${parsed.type}`);
  }

  return parsed as ServerMessage;
}

function isObject(value: unknown): value is { readonly type?: unknown } {
  return typeof value === "object" && value !== null;
}
